### PR TITLE
fix(typescript): fix proxyRes and router types

### LINKED ISSUE #407: typing for proxyRes seems to be incorrect
### Steps to reproduce

Using TypeScript, this is valid but throws an error.
```ts
// options
onProxyRes: (proxyRes, req, res) => {
  const incomingCookies = proxyRes.getHeader(SET_COOKIE_HEADER);
  // etc
},
```

### Expected behavior

* TypeScript doesn't expect `proxyRes.getHeader`, `proxyRes.hasHeader`, `proxyRes.setHeader` to be methods available.
* TypeScript expects `proxyRes.headers` to be an array of strings.

### Actual behavior

```bash
 ✖  uncaughtException (an exception was thrown but not caught)  TypeError: proxyRes.getHeader is not a function
    at ProxyServer.onProxyRes (proxy.ts)
````

### Workaround

`proxyRes.headers` works as expected, so adding this code works, but it would be cleaner to just have the correct types.

```ts
// options
onProxyRes: (
  proxyRes: ServerResponse & { headers: { [name: string]: string[] } },
  req,
  res
) => {
  const incomingCookies = proxyRes.headers[SET_COOKIE_HEADER];
  // etc
},

```

I'm not sure what type the `proxyRes` should be.

#### client info

I'm using latest Express.

### Reproducible Demo

I can create one if necessary.

### LINKED ISSUE #408: typing and documentation for router's return value seems to be incorrect
### Steps to reproduce

Using TypeScript:

This is typed correctly, but does not work:
```ts
// options
router: (req) => {
	const newTarget = { 
		host: req.context.host || 'localhost',
		protocol: req.context.protocol || 'https:',
		port: req.context.port || 8080
	};
	return `${target.protocol}//${host}:${port}`;
},
```

The return value is typed incorrectly, but this works:

```ts
// options
router: (req) => {
	const newTarget = { 
		host: req.context.host || 'localhost',
		protocol: req.context.protocol || 'https:',
		port: req.context.port || 8080
	};

	return newTarget;
},
```

### Expected behavior

* `router` return type seems like it should be something like:
```ts
interface RouterReturnValue { 
	host: string; 
	protocol: 'http:' | 'https:';  // the : is required!
	port: number;
}
```

### Actual behavior

Using a string doesn't work to dynamically set the `target`.

### Workaround

```ts
router: (req) => {
	const newTarget = { 
		host: req.context.host || 'localhost',
		protocol: req.context.protocol || 'https:',
		port: req.context.port || 8080
	};
	return (newTarget as unknown) as string;
}
```

### Reproducible Demo

I'm not sure yet how to make a standalone example but I can try if needed.

### PR DESCRIPTION
**Pull request by dylang**
* Fixes #407 
* Fixes #408 
* Explicit about `protocol` requiring `:` in README.
* Adds `https:` to router tests.
* Test `protocol: 'https'` to router tests showing `:` is required for `https` to work.

### CODE REVIEW

**Comment by dylang** [comment]
Hi @chimurai, any changes you want to this?

File: README.md [hunk 1]
```diff
@@ -245,7 +244,11 @@ Providing an alternative way to decide which requests should be proxied; In case
 
   // Custom router function
   router: function(req) {
-      return 'http://localhost:8004';
+      return {
```
Comment (chimurai) [inline README.md]: Could keep the original example?

Basically providing both examples; One with target as a string and one with the target as an object.

**Comment by chimurai** [comment]
Great work! Just one comment on the documentation.

Otherwise LGTM 👍

File: README.md (hunk 1 above)
Comment (dylang) [reply README.md]: The example under this one is still the same, do you think it's enough?

```
// Asynchronous router function which returns promise
  router: async function(req) {
      const url = await doSomeIO();
      return url;
  }
```

Comment (chimurai) [reply README.md]: Think splitting the original example to a string and object example would make more sense.
It'll give new user just examples to copy/paste from.

```javascript
// Custom router function (string target)
router: function(req) {
    return 'http://localhost:8004';
}
```

```javascript
// Custom router function (target object)
router: function(req) {
    return {
      protocol: 'https:', // The : is required
      host: 'localhost',
      port: 8004
    };
}
```

The async example just indicates router accepts a Promise. Hopefully devs can figure out you can resolve the Promise with the same string/object `target`

Comment (dylang) [reply README.md]: Good idea! I made the change.

**Comment by chimurai** [comment]
published v1.0.1 to npm

Thanks for helping out with the types!

**Pull request merged**